feat: - avm/res/logic/integration-account

lsnoddy opened the pull request:
## Description

This pull request introduces a new Bicep module for managing Azure Logic Integration Accounts, adding support for agreements, assemblies, certificates, maps, partners, and schemas. 


Closes https://github.com/Azure/Azure-Verified-Modules/issues/2016


-->

## Pipeline Reference

<!-- Insert your Pipeline Status Badge below -->

| Pipeline |
| -------- |
[![avm.res.logic.integration-account](https://github.com/lsnoddy/bicep-registry-modules/actions/workflows/avm.res.logic.integration-account.yml/badge.svg?branch=users%2Flsnoddy%2Favm-res-logic-integration-account)](https://github.com/lsnoddy/bicep-registry-modules/actions/workflows/avm.res.logic.integration-account.yml)

## Type of Change

<!-- Use the checkboxes [x] on the options that are relevant. -->

- Azure Verified Module updates:
  - [x] New Module
  - [ ] Bugfix containing backwards-compatible bug fixes, and I have NOT bumped the MAJOR or MINOR version in `version.json`:
  - [ ] Feature update backwards compatible feature updates, and I have bumped the MINOR version in `version.json`.
  - [ ] Breaking changes and I have bumped the MAJOR version in `version.json`.
  - [ ] Update to documentation
- [ ] Update to CI Environment or utilities (Non-module affecting changes)

## Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [x] I have run `Set-AVMModule` locally to generate the supporting module files.
- [x] My corresponding pipelines / checks run clean and green without any errors or warnings
- [x] I have updated the module's CHANGELOG.md file with an entry for the next version

<!--  Please keep up to date with the contribution guide at https://aka.ms/avm/contribute/bicep -->

AlexanderSehr commented on avm/res/logic/integration-account/agreement/main.bicep:
```diff
@@ -0,0 +1,85 @@
+param guestPartner string
+
+@description('Required. The business identity of the host partner.')
+param hostIdentity businessIdentityType
```
Why a custom and not a resource-derived type if only the standard properties are used?

lsnoddy replied:
Changed to resource-derived type

AlexanderSehr commented on avm/res/logic/integration-account/certificate/main.bicep:
```diff
@@ -0,0 +1,74 @@
+param integrationAccountName string
+
+@description('Optional. The key details in the key vault.')
+param key keyVaultKeyReferenceType?
```
Why not use a resouce derived type `resourceInput<'Microsoft.Logic/integrationAccounts/certificates@2019-05-01'>.properties.key` ? :)

AlexanderSehr commented on avm/res/logic/integration-account/map/main.bicep:
```diff
@@ -0,0 +1,62 @@
+param mapType ('Liquid' | 'NotSpecified' | 'Xslt' | 'Xslt20' | 'Xslt30') = 'Xslt'
+
+@description('Optional. The map metadata.')
+param metadata object?
```
```suggestion
param metadata resourceInput<'Microsoft.Logic/integrationAccounts/maps@2019-05-01'>.properties.metadata?
```

AlexanderSehr commented on avm/res/logic/integration-account/map/main.bicep:
```diff
@@ -0,0 +1,62 @@
+param metadata object?
+
+@description('Optional. The parameters schema of integration account map.')
+param parametersSchema object?
```
```suggestion
param parametersSchema resourceInput<'Microsoft.Logic/integrationAccounts/maps@2019-05-01'>.properties.parametersSchema?
```

AlexanderSehr commented on avm/res/logic/integration-account/partner/main.bicep:
```diff
@@ -0,0 +1,98 @@
+param integrationAccountName string
+
+@description('Optional. B2B partner content settings.')
+param b2bPartnerContent b2bPartnerContentType?
```
Can use the resource-derived type
```suggestion
param b2bPartnerContent resourceInput<'Microsoft.Logic/integrationAccounts/partners@2019-05-01'>.properties.b2bPartnerContent?
```

lsnoddy replied:
Changed to derived type and matched api spec with resourceInput<'Microsoft.Logic/integrationAccounts/partners@2019-05-01'>.properties.content.b2b?

AlexanderSehr commented on avm/res/logic/integration-account/schema/main.bicep:
```diff
@@ -0,0 +1,64 @@
+param documentName string?
+
+@description('Optional. The schema metadata.')
+param metadata object?
```
```suggestion
param metadata resourceInput<'Microsoft.Logic/integrationAccounts/schemas@2019-05-01'>.properties.metadata?
```

AlexanderSehr commented on avm/res/logic/integration-account/tests/e2e/max/main.test.bicep:
```diff
@@ -0,0 +1,483 @@
+  name: '${uniqueString(deployment().name, resourceLocation)}-nestedDependencies'
+  params: {
+    managedIdentityName: 'dep-${namePrefix}-msi-${serviceShort}'
+    location: resourceLocation
```
```suggestion
```

AlexanderSehr commented on avm/res/logic/integration-account/tests/e2e/waf-aligned/main.test.bicep:
```diff
@@ -0,0 +1,211 @@
+    logAnalyticsWorkspaceName: 'dep-${namePrefix}-law-${serviceShort}'
+    eventHubNamespaceEventHubName: 'dep-${namePrefix}-evh-${serviceShort}01'
+    eventHubNamespaceName: 'dep-${namePrefix}-evhns-${serviceShort}01'
+    location: resourceLocation
```
```suggestion
```

AlexanderSehr commented on avm/res/logic/integration-account/main.bicep:
```diff
@@ -0,0 +1,511 @@
+@description('Optional. The lock settings of the service.')
+param lock lockType?
+
+import { roleAssignmentType } from 'br/public:avm/utl/types/avm-common-types:0.2.1'
```
```suggestion
import { roleAssignmentType } from 'br/public:avm/utl/types/avm-common-types:0.6.1'
```

AlexanderSehr commented on avm/res/logic/integration-account/main.bicep:
```diff
@@ -0,0 +1,511 @@
+  b2bPartnerContent: b2bPartnerContentType?
+
+  @description('Optional. The partner metadata.')
+  metadata: object?
```
This and other metadata references should be replaced with the resource derived type I commented in the individual child-module files

AlexanderSehr replied:
Hey @lsnoddy,
thanks for the comment. That is ... something. I'll try to reproduce this and find a solution

AlexanderSehr replied:
Alright, @lsnoddy, I was able to replicate the issue myself. The best solution I was able to come up with which seems to work
- Use Resource-Dervied-Types for all 'metadata' parameters but for the modules
  - partners
  - schemas
  - maps

For the parameter, use instead
```bicep
@description('Optional. The metadata.')
param metadata {
  @description('Optional. A metadata key-value pair.')
  *: string?
}? // Resource-derived type fails PSRule
```

for their types in the parent module 'integration-account', instead use
```bicep
  @description('Optional. The metadata.')
  metadata: {
    @description('Optional. A metadata key-value pair.')
    *: string?
  }? // Resource-derived type fails PSRule
```

AlexanderSehr commented:
Hey @lsnoddy,
thank you for the great PR and applying all these changes :) Please note that there are very few remaining that don't show up in the feed above until you expand the collapsed sections (multiple times). The feed is too smart for its own good 😄

lsnoddy commented on avm/res/logic/integration-account/tests/e2e/max/main.test.bicep:
```diff
@@ -0,0 +1,483 @@
+  </xsl:template>
+</xsl:stylesheet>'''
+
+var assemblyContent = 'TVqQAAMAAAAEAAAA//8AALgAAAAAAAAAQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAgAAAAA4fug4AtAnNIbgBTM0hVGhpcyBwcm9ncmFtIGNhbm5vdCBiZSBydW4gaW4gRE9TIG1vZGUuDQ0KJAAAAAAAAABQRQAATAEDAIIBUIMAAAAAAAAAAOAAIgALATAAAAgAAAAIAAAAAAAAIicAAAAgAAAAQAAAAABAAAAgAAAAAgAABAAAAAAAAAAEAAAAAAAAAACAAAAAAgAAAAAAAAMAYIUAABAAABAAAAAAEAAAEAAAAAAAABAAAAAAAAAAAAAAAM4mAABPAAAAAEAAAHwFAAAAAAAAAAAAAAAA ...
```
Changed to load data from files for each test case

lsnoddy commented on avm/res/logic/integration-account/main.bicep:
```diff
@@ -0,0 +1,511 @@
+  tags: object?
+}
+
+import { businessIdentityType } from 'agreement/main.bicep'
```
Replaced with resource-derived type

lsnoddy commented:
@AlexanderSehr, thank you for going through all of this and providing your feedback. The only thing I was not able to implement was the resource derived type for metadata, as the waf-aligned psrule jobs failed when it was implemented. I added a screenshot of the pipeline error above and changed metadata back to using object? for the child modules that get used in the waf test case.

lsnoddy commented on avm/res/logic/integration-account/main.bicep:
```diff
@@ -0,0 +1,511 @@
+@description('Optional. Location for all Resources.')
+param location string = resourceGroup().location
+
+import { diagnosticSettingFullType } from 'br/public:avm/utl/types/avm-common-types:0.2.1'
```
This has been updated to 0.6.1

AlexanderSehr commented on avm/res/logic/integration-account/main.bicep:
```diff
@@ -0,0 +1,508 @@
+  documentName: string?
+
+  @description('Optional. The schema metadata.')
+  metadata: object?
```
Should use resource-derived-type

AlexanderSehr commented on avm/res/logic/integration-account/main.bicep:
```diff
@@ -0,0 +1,508 @@
+  parametersSchema: object?
+
+  @description('Optional. The map metadata.')
+  metadata: object?
```
Should use resource-derived-type

AlexanderSehr commented on avm/res/logic/integration-account/main.bicep:
```diff
@@ -0,0 +1,508 @@
+  hostPartner: string
+
+  @description('Required. The agreement content settings.')
+  content: object
```
Should have type

lsnoddy replied:
Changed to resource derived type
